app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/WifiDirectManager.java

```diff
@@ -0,0 +1,481 @@
+    }
+
+    manager = (WifiP2pManager) activity.getSystemService(Context.WIFI_P2P_SERVICE);
+    channel = manager.initialize(activity, getMainLooper(), null);
+    registerWifiDirectBroadcastReceiver();
+  }
+
+  public void registerWifiDirectBroadcastReceiver() {
+    receiver = new KiwixWifiP2pBroadcastReceiver(this);
+
+    // For specifying broadcasts (of the P2P API) that the module needs to respond to
+    IntentFilter intentFilter = new IntentFilter();
+    intentFilter.addAction(WifiP2pManager.WIFI_P2P_STATE_CHANGED_ACTION);
+    intentFilter.addAction(WifiP2pManager.WIFI_P2P_PEERS_CHANGED_ACTION);
+    intentFilter.addAction(WifiP2pManager.WIFI_P2P_CONNECTION_CHANGED_ACTION);
+    intentFilter.addAction(WifiP2pManager.WIFI_P2P_THIS_DEVICE_CHANGED_ACTION);
+
+    activity.registerReceiver(receiver, intentFilter);
+  }
+
+  public void unregisterWifiDirectBroadcastReceiver() {
+    activity.unregisterReceiver(receiver);
+  }
+
+  public void discoverPeerDevices() {
+    manager.discoverPeers(channel, new WifiP2pManager.ActionListener() {
+      @Override
+      public void onSuccess() {
+        displayToast(R.string.discovery_initiated, Toast.LENGTH_SHORT);
+      }
+
+      @Override
+      public void onFailure(int reason) {
+        String errorMessage = getErrorMessage(reason);
+        Log.d(TAG, activity.getString(R.string.discovery_failed) + ": " + errorMessage);
+        displayToast(R.string.discovery_failed, Toast.LENGTH_SHORT);
+      }
+    });
+  }
+
+  @Override
+  public void onWifiP2pStateChanged(boolean isEnabled) {
+    this.isWifiP2pEnabled = isEnabled;
+
+    if(!isWifiP2pEnabled) {
+      displayToast(R.string.discovery_needs_wifi, Toast.LENGTH_SHORT);
+      callbacks.onConnectionToPeersLost();
+    }
+
+    Log.d(TAG, "WiFi P2P state changed - " + isWifiP2pEnabled);
+  }
+
+  @Override
+  public void onPeersChanged() {
+    /* List of available peers has changed, so request & use the new list through
+     * PeerListListener.requestPeers() callback */
+    manager.requestPeers(channel, this);
+    Log.d(TAG, "P2P peers changed");
+  }
+
+  @Override
+  public void onConnectionChanged(boolean isConnected) {
+    if (isConnected) {
+      // Request connection info about the wifi p2p group formed upon connection
+      manager.requestConnectionInfo(channel, this);
+    } else {
+      // Not connected after connection change -> Disconnected
+      callbacks.onConnectionToPeersLost();
+    }
+  }
+
+  @Override
+  public void onDeviceChanged(@Nullable WifiP2pDevice userDevice) {
+    // Update UI with wifi-direct details about the user device
+    callbacks.onUserDeviceDetailsAvailable(userDevice);
+  }
+
+  public boolean isWifiP2pEnabled() {
+    return isWifiP2pEnabled;
+  }
+
+  /* From WifiP2pManager.ChannelListener interface */
+  @Override
+  public void onChannelDisconnected() {
+    // Upon disconnection, retry one more time
+    if (!retryChannel) {
```

Comment:
change the name of this boolean to `hasRetriedAlready` or invert it and call it `shouldRetry`.
Right now `Not Retry Channel` actually retries the channel which is a wee bit odd

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/e0651dd88c2ee30b8e17d54ff215384a1d877681#diff-7e01f6f425daaca67175b74c839c6af2R57)